patroni/quorum.py

```diff
@@ -0,0 +1,429 @@
+
+        safety_margin = self.quorum + min(self.numsync, self.numsync_confirmed) - len(self.voters | self.sync)
+        if safety_margin > 0:  # In the middle of changing replication factor.
+            if self.numsync > self.sync_wanted:
+                logger.debug('Case 3: replication factor is bigger than needed')
+                yield from self.sync_update(max(self.sync_wanted, len(self.voters) - self.quorum), self.sync)
+            else:
+                logger.debug('Case 4: quorum is bigger than needed')
+                yield from self.quorum_update(len(self.sync) - self.numsync, self.voters)
+        else:
+            safety_margin = self.quorum + self.numsync - len(self.voters | self.sync)
+            if self.numsync == self.sync_wanted and safety_margin > 0 and self.numsync > self.numsync_confirmed:
+                yield from self.quorum_update(len(self.sync) - self.numsync, self.voters)
+
+    def __remove_gone_nodes(self) -> Iterator[Transition]:
+        """Remove inactive nodes from ``synchronous_standby_names`` and from ``/sync`` key.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        to_remove = self.sync - self.active
+        if to_remove and self.sync == to_remove:
+            logger.debug("Removing nodes: %s", to_remove)
+            yield from self.quorum_update(0, CaseInsensitiveSet(), adjust_quorum=False)
+            yield from self.sync_update(0, CaseInsensitiveSet())
+        elif to_remove:
+            logger.debug("Removing nodes: %s", to_remove)
+            can_reduce_quorum_by = self.quorum
+            # If we can reduce quorum size try to do so first
+            if can_reduce_quorum_by:
+                # Pick nodes to remove by sorted order to provide deterministic behavior for tests
+                remove = CaseInsensitiveSet(sorted(to_remove, reverse=True)[:can_reduce_quorum_by])
+                sync = CaseInsensitiveSet(self.sync - remove)
+                # when removing nodes from sync we can safely increase numsync if requested
+                numsync = min(self.sync_wanted, len(sync)) if self.sync_wanted > self.numsync else self.numsync
+                yield from self.sync_update(numsync, sync)
+                voters = CaseInsensitiveSet(self.voters - remove)
+                to_remove &= self.sync
+                yield from self.quorum_update(len(voters) - self.numsync, voters,
+                                              adjust_quorum=not to_remove)
+            if to_remove:
+                assert self.quorum == 0
+                numsync = self.numsync - len(to_remove)
+                sync = CaseInsensitiveSet(self.sync - to_remove)
+                voters = CaseInsensitiveSet(self.voters - to_remove)
+                sync_decrease = numsync - min(self.sync_wanted, len(sync))
+                quorum = min(sync_decrease, len(voters) - 1) if sync_decrease else 0
+                yield from self.quorum_update(quorum, voters, adjust_quorum=False)
```

Comment:
I must say I'm having some issues to consume the maths shown in this module.

I wonder if is there any way in which we can improve readability/consumption of the logic that has been implemented?

Maybe we could:

* Expand docstrings
* Expand comments
* Create auxiliary variables with a detailed description
* Describe the workflow

I don't know. I wonder if it's just me who is having bit of a hard time to digest it all?

Reply:
(I put a localized comment here, but I guess we could look for improvements in other pieces of this module)

Reply:
Yes, I know, it is extremely complex. The original PR was opened more than 5 years ago and for me it took a few attempts to get the idea.
All the complexity exists because we can't atomically update both, `/sync`  and `synchronous_standby_names`.

I would recommend you to start understanding the workflow from the `__add_new_nodes()` method:
- 'a' is a leader
- `/sync` key is empty
- `synchronous_standby_names` is empty
- we have 'b' and 'c' nodes streaming

The `QuorumStateResolver` will produce following transitions:
1. 'sync', 'a', 1, {'b', 'c'}
2. 'restart', because after updating `synchronous_standby_names` we need to wait until 'b' and 'c' will catch up with the LSN when `postgresql.conf` was updated.

That is, we got to one of **non-steady cases,** when  `/sync`  and `synchronous_standby_names` don't match.
Such state will remain until at least one of the nodes 'b' or 'c' will catch up with the LSN when `postgresql.conf` was updated (`numsync_confirmed>0`).


Interrupt might happen not only because of `restart` yielded by  `QuorumStateResolver`, but also because of other external factors like for example not being able to update the `/sync` key.
Lets consider the state:
- `a` is a leader
- `/sync` key: {leader: a, sync_standby: b,c, quorum: 1}
- `synchronous_standby_names`: `{ANY 1 (b, c)}`
- only `b` is streaming

In this case `QuorumStateResolver` will produce following transitions:
1. 'sync', 'a', 1, {'b'}
2. 'quorum', 'a', 0, {'b'}

That is, first 'c' is removed from the `synchronous_standby_names` and only after from the '/sync' key.
If update of the `/sync` key fails, we also get to a **non-steady** state.

Non-steady state might also happen when replication factor (`sync_wanted`) is increased/reduced by the administrator. The change also must be reflected by performing updates of `/sync`  and `synchronous_standby_names` in a very specific order.

Last, but not least, there could also be a combination of different changes. Like for example new nodes started streaming, but admin decided to increase/decrease replication factor.

If you look at unit-tests, you'll see that `QuorumStateResolver` can produce up to 4 transitions until the steady state is reached: https://github.com/zalando/patroni/blob/feature/quorum-commit/tests/test_quorum.py#L410-L422 
And, `QuorumStateResolver` must be able to "resume" the chain from the arbitrary state when it was interrupted.
Most of this magic happens in `__handle_non_steady_cases()` method and this is why it is com complex.

Unit-tests actually check restart `QuorumStateResolver` from all intermediate states and check that yielded transitions are matching the expectation: https://github.com/zalando/patroni/blob/feature/quorum-commit/tests/test_quorum.py#L10-L28